### ISSUE TITLE
How best to change item in subModelSeq

### ISSUE THREAD

**Issue by MicaelMor** [elmish/Elmish.WPF#305]
Hi,

I have a question regarding "best practices" for doing something.
Say I have a model with a list of a certain record type, and I want to change an element in that list, currently I have the update happening this way in the message (names and model simplified for example’ sake):

```fsharp
|UpdateSomeField x ->
        let newList =
                m.Entities
                |> List.map (fun entity ->
                        if entity.Id = m.SelectedEntityId then {entity with SomeField = x}
                        else entity
        {m with Entities = newList}, Cmd.none
```
and the bindings are like this:

```fsharp
"Entities"
|> Binding.subModelSeq
       ((fun m -> m.Entities),
       (fun e -> e.Id),
       (fun () ->
              [ "SomeField |> Binding.twoWay ((fun (_, e) -> e.SomeField), (fun newVal _ -> UpdateSomeField newVal))]))

"SelectedEntityId" |> Binding.subModelSelectedItem("Entities", (fun m -> m.SelectedEntityId), SelectEntityId)
```

Outside of creating a function that takes a model and a new entity to reduce on the boilerplate for each update field. Is the better way of doing this to use the more complex subModelSeq signature where the subModel gets its own model and update and bindings, or is there some other way that I am missing?
And does using the more complex subModelSeq even make a difference, like say does it do anything different under the hood to make it more optimized for large lists and so on?

Thank you.

**Comment by TysonMN** [comment]
First, here is how you can better format your code on GitHub.  Quote this comment to see the syntax.
```fsharp
| UpdateSomeField x ->
  let newList =
    m.Entities
    |> List.map (fun entity -> if entity.Id = m.SelectedEntityId
                               then {entity with SomeField = x}
                               else entity {m with Entities = newList}, Cmd.none)
```

```fsharp
"Entities" |> Binding.subModelSeq(
  (fun m -> m.Entities),
  (fun e -> e.Id),
  (fun () ->
    [ "SomeField |> Binding.twoWay ((fun (_, e) -> e.SomeField), (fun newVal _ -> UpdateSomeField newVal))]))

"SelectedEntityId" |> Binding.subModelSelectedItem("Entities", (fun m -> m.SelectedEntityId), SelectEntityId)
```

> Outside of creating a function that takes a model and a new entity to reduce on the boilerplate for each update field. Is the better way of doing this to use the more complex subModelSeq signature where the subModel gets its own model and update and bindings, or is there some other way that I am missing?

In general, yes, I think it is better to use [the most complex of the three `Binding.subModelSeq` overloads](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1740-L1746).  In my application at work, I think I only use that overload.

Specifically though, I am confused about something in your example.  The `UpdateSomeField` message is sent from a sub-binding of a `SubModelSeq` binding but it is applied to the selected item of the sequence.  Having the message sent from the sub-binding makes me think that all items in the sequence can be edited it the same way (like how the every counter can be edited in the `SubModelSeq` sample).  On the other hand, applying the message to the selected item makes me think the your have a master-detail arrangement and the editing is occuring in the detail view of the selected item.

Can you clear up my confusing by share an executable example?

> And does using the more complex subModelSeq even make a difference, like say does it do anything different under the hood to make it more optimized for large lists and so on?

No, the performance is essentially the same among the three `Binding.subModelSeq` overloads.  Your update code has a much bigger impact on the performance.

**Comment by MicaelMor** [comment]
Thanks for the formatting tip, have updated my original post so it reads a lot better.

As for your question I believe the confusion comes from a misinterpretation of the `UpdateSomeField` message due to the poor original formatting, when you formatted it better you changed the code to something that would have a model similar to the counter in `SubModelSeq` sample, where the subModel is recursive, while the model in my example is something much simpler like this:

```fsharp
type SomeRecord = {
    Id: int
    SomeField: string
    AnotherField: float}

type Model = {
    Entities: SomeRecord list
    SelectedEntityId: int
}
```

(Added `AnotherField` to make it clear this record would have other fields)

All I'm trying to do with this model is to change the fields in one of the elements of the `Entities` (the one that matches the `Id` in `SelectedEntityId` ofc).
For a more concrete example of usage, the user selects a `SomeRecord` in a grid, and then changes the fields directly in the grid.
If there is still some confusion please let me know.

> No, the performance is essentially the same among the three Binding.subModelSeq overloads. Your update code has a much bigger impact on the performance.

So to clarify any of the Binding.subModelSeq are going to be more performant than the Update code I displayed above correct?

**Comment by TysonMN** [comment]
(Since you improved for your formatting, I deleted my formatting tip.  I should have put that information in a separate comment and now marked it as resolved.)

> If there is still some confusion please let me know.
> 
> https://i.imgur.com/sMl3dBJ.png <- basically this.
> 
> EDITED: To clarify the reason why this needs to be a submodel instead of just having the `SomeRecord` fields in the base model, is because it is indeed a master detail as you said, where the contents of the `Entities` (`SomeRecord list`) depends on some field selected previously, something that I didn't make clear previously because didn't consider it relevant to the update question.

Mostly the same confusion.

You say that "it is indeed a master detail", but I wouldn't say the linked image is a master-detail interface.  [Here is an example of a master-detail interface](https://github.com/tjvantoll/angular-master-detail).  The point is that you can only see the details of the selected item.  In contrast, the image to which you linked shows the columns (i.e. details) for every value in the (master) list.  You could maintain your editing behavior without having a selected item and the resulting editing code would be simpler.  (Of course not having a selected item would be simpler, but I am saying that code just to edit `SomeField` for some item would be simpler.)

It would be great if you could share an executable example.

> So to clarify any of the `Binding.subModelSeq` are going to be more performant than my code where I replace the entire list with a new list correct?

I don't exactly understand what is being compared.  Performance optimization can be a tricky thing.

The general wisdom is to delay performance optimizations until the client / product owner / users complain that it is slow.  I generally believe this.  At the same time, I think this wisdom is incorrect to the extent that decisions to delay performance optimizations were accompanied by decisions to design the code so that performance optimizations are prohibitively difficult to implement.

Thankfully, I believe that functional programming is a good general strategy for making code design decisions that preserve the ability to make relatively easy performance optimizations.

**Comment by MicaelMor** [comment]
Hi,

> Mostly the same confusion.
> 
> You say that "it is indeed a master detail", but I wouldn't say the linked image is a master-detail interface. Here is an example of a master-detail interface. The point is that you can only see the details of the selected item. In contrast, the image to which you linked shows the columns (i.e. details) for every value in the (master) list. You could maintain your editing behavior without having a selected item and the resulting editing code would be simpler. (Of course not having a selected item would be simpler, but I am saying that code just to edit SomeField for some item would be simpler.)
> 
> It would be great if you could share an executable example.

Thanks for the reply, have created a sample based on a fork of V4 on the project SubModelSeq https://github.com/MicaelMor/Elmish.WPF/tree/subModelSeq_Question_Sample

And here is a full picture of the real software https://i.imgur.com/shCWN0E.png can't really show records or I would have to blur out a fair bit of it (haven't checked how to use design time VM yet), but it is quite simple, the software works from top to bottom, with for the most part each subsequent field working with information of the previous field (the exception are the 2 combo boxes, but they work in a similar fashion) the steps of the software are as follows:

1. Select Cliente from the dropdown list field (this is 2 labels in the model, 1 SelectedCliente of type Cliente another ClienteList of type Cliente list)
2. Morada  dropdown gets filtered based on SelectedCliente and then the user here selects a Morada (this is 2 labels in the model, 1 SelectedMorada of type Morada another MoradaList of type Morada list)
3. When Morada is selected the first Grid gets updated with the records based on the SelectedMorada, this grid allows selection of record but has no update of the fields in it, it is purely read only (same as before this is 2 labels in the model, 1 SelectedMovimento of type Movimento another MovimentoList of type Movimento list)
4. The following 2 combo boxes have the same pattern as all the previous ones, where each has a selected type and a list of the same type as the selected one
5. First button `Adicionar Movimento` takes all the previously selected information and creates a new record that gets added to the grid below.
6. This grid which is the one I had previously linked an image, has those 4 labels all of which need to be updatable (this is 2 labels in the model, 1 SelectedFatura of type Fatura another FaturaList of type Fatura list)
7. Then the final button grabs previously provided information (SelectedFatura is only relevant for Changing fields in this project) and does its thing, at the end of it, it resets the model.

Hope this clears the confusion, and if you have any tips on how the sample should be changed, I would appreciate it, since while obviously a simpler example than the real project, the idea is the same, which is to modify values held in a list of record types through a wpf grid.

> I don't exactly understand what is being compared. Performance optimization can be a tricky thing.
> 
> The general wisdom is to delay performance optimizations until the client / product owner / users complain that it is slow. I generally believe this. At the same time, I think this wisdom is incorrect to the extent that decisions to delay performance optimizations were accompanied by decisions to design the code so that performance optimizations are prohibitively difficult to implement.
> 
> Thankfully, I believe that functional programming is a good general strategy for making code design decisions that preserve the ability to make relatively easy performance optimizations.

Shouldn't have asked about performance, since that is indeed entirely dependent on the project, and for my particular current project it is entirely a non issue since we are talking about a list that is likely to at best end up having 10 records at a time, so one could create new lists dozens of times in a row and still have it be virtually instantaneous for the user.

Was mostly asking about performance as a way to know if subModelSeq does anything more fancy than just creating an entirely new list (aka what my sample code is doing), so as to keep in mind for other projects where this might be a concern, I assume it does.

Agree have found functional programming to be very good for significant refactoring without breaking the code, since most invalid states are checked at compile time, if the program compiles after refactoring code, it is usually working as intended, and if not there is a good chance the problem was already there, this ofc assuming one didn't invalidate compile time checks with shortcuts by not using Result or not checking option values before using them and so on.

Have ended up writing far too large portions of code without testing only to have the code work flawlessly, or failing only due to WPF bindings, or other external non type provider backed sources.

**Comment by TysonMN** [comment]
> haven't checked how to use design time VM yet

You really should.  [Here](https://github.com/elmish/Elmish.WPF#can-i-use-design-time-view-models) are instructions for how to add design-time support.  Also, every sample is using design-time view models.

**Comment by TysonMN** [comment]
> I assume when you say you can change the records without a Selected field, that is accomplished using the more complex `subModelSeq` correct?

Correct.  Using [either of the two more](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1740-L1746) [complex overloads of `Binding.subModelSeq`](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1771-L1776) includes the argument
```fsharp
toMsg: 'id * 'subMsg -> 'msg
```

You can use that `'id` instead of the ID of the selected item.

**Comment by TysonMN** [comment]
> Was mostly asking about performance as a way to know if `subModelSeq` does anything more fancy than just creating an entirely new list (aka what my sample code is doing), so as to keep in mind for other projects where this might be a concern, I assume it does.

Oh, yes; it is very efficient.

A `SubModelSeq` binding stores an `ObservableCollection` for every item in the given sequence.  When [the binding is updated](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/ViewModel.fs#L680-L688), it mutates that collection to match the more recently given sequence.  Generally speaking, I know this algorithmic problem by the name `merge`.  [The merge algorithm used](https://github.com/elmish/Elmish.WPF/blob/master/src/Elmish.WPF/ViewModel.fs#L16-L178) is essentially the same as in the Elm programming language.

This algorithm is optimized to take the least amount of time when the previous sequence and the new sequence are nearly the same (such as a single addition, a single removal, and a single swap).  It also takes the least amount of time when there are multiple such changes but they do not occur sequentially.  In contrast, when the new sequence has been filtered or sorted, then essentially all existing items are removed and all new items are created and added.

**Comment by TysonMN** [comment]
> if you have any tips on how the sample should be changed, I would appreciate it, since while obviously a simpler example than the real project, the idea is the same, which is to modify values held in a list of record types through a wpf grid.

I will look at this tomorrow.

**Comment by TysonMN** [comment]
I just created PR https://github.com/MicaelMor/Elmish.WPF/pull/1.  I think it solves your problem.

> Was mostly asking about performance as a way to know if `subModelSeq` does anything more fancy than just creating an entirely new list (aka what my sample code is doing), so as to keep in mind for other projects where this might be a concern, I assume it does.
> ...
> ...so one could create new lists dozens of times in a row and still have it be virtually instantaneous for the user.

The only extra cost is the time spent in your code creating dozens of lists.  The performance of `SubModelSeq` is the same as if you provided the same list as last time.  (A future version might optimize that case more via a reference equality check, but that is not happening now.)

I didn't know what you meant by your code creating an entirely new list.  This is why your changes to `SomeField` were not being saved.  In my PR, I changed your model so store all the entities and the binding to only be giving the filtered ones.

Does that solve your problem?

**Comment by MicaelMor** [comment]
Sorry for not answering before but was deploying the project so things got a bit busy the last few days, anyway as all the questions I had have been answered I will close this issue now.

Thanks for all the help and information.

> > haven't checked how to use design time VM yet
> 
> You really should. [Here](https://github.com/elmish/Elmish.WPF#can-i-use-design-time-view-models) are instructions for how to add design-time support. Also, every sample is using design-time view models.

Thanks for the link to the documentation, it was quite simple to implement, pretty much just a copy and paste job,

> > I assume when you say you can change the records without a Selected field, that is accomplished using the more complex `subModelSeq` correct?
> 
> Correct. Using [either of the two more](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1740-L1746) [complex overloads of `Binding.subModelSeq`](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1771-L1776) includes the argument
> 
> ```fsharp
> toMsg: 'id * 'subMsg -> 'msg
> ```
> 
> You can use that `'id` instead of the ID of the selected item.

Ok, thanks for the confirmation.

> > Was mostly asking about performance as a way to know if `subModelSeq` does anything more fancy than just creating an entirely new list (aka what my sample code is doing), so as to keep in mind for other projects where this might be a concern, I assume it does.
> 
> Oh, yes; it is very efficient.
> 
> A `SubModelSeq` binding stores an `ObservableCollection` for every item in the given sequence. When [the binding is updated](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/ViewModel.fs#L680-L688), it mutates that collection to match the more recently given sequence. Generally speaking, I know this algorithmic problem by the name `merge`. [The merge algorithm used](https://github.com/elmish/Elmish.WPF/blob/master/src/Elmish.WPF/ViewModel.fs#L16-L178) is essentially the same as in the Elm programming language.
> 
> This algorithm is optimized to take the least amount of time when the previous sequence and the new sequence are nearly the same (such as a single addition, a single removal, and a single swap). It also takes the least amount of time when there are multiple such changes but they do not occur sequentially. In contrast, when the new sequence has been filtered or sorted, then essentially all existing items are removed and all new items are created and added.

This was exactly the kind of information I was looking for, thank you.

> I just created PR [MicaelMor#1](https://github.com/MicaelMor/Elmish.WPF/pull/1). I think it solves your problem.
> 
> > Was mostly asking about performance as a way to know if `subModelSeq` does anything more fancy than just creating an entirely new list (aka what my sample code is doing), so as to keep in mind for other projects where this might be a concern, I assume it does.
> > ...
> > ...so one could create new lists dozens of times in a row and still have it be virtually instantaneous for the user.
> 
> The only extra cost is the time spent in your code creating dozens of lists. The performance of `SubModelSeq` is the same as if you provided the same list as last time. (A future version might optimize that case more via a reference equality check, but that is not happening now.)
> 
> I didn't know what you meant by your code creating an entirely new list. This is why your changes to `SomeField` were not being saved. In my PR, I changed your model so store all the entities and the binding to only be giving the filtered ones.
> 
> Does that solve your problem?

Yes, thank you for the modifications on the sample project.

**Comment by TysonMN** [comment]
> In general, yes, I think it is better to use [the most complex of the three `Binding.subModelSeq` overloads](https://github.com/elmish/Elmish.WPF/blob/2d80782826606a0addc9313eecce4099227b8f88/src/Elmish.WPF/Binding.fs#L1740-L1746). In my application at work, I think I only use that overload.

Since I wrote that, I have simplified the `SubModelSeq` API.  Now this overload exists in `master`, which is the simplest.

https://github.com/elmish/Elmish.WPF/blob/c10be058b14a4049ecfd8843f497bdfec1cf2c4a/src/Elmish.WPF/Binding.fs#L2329-L2331

And this is the next simplest overload.
https://github.com/elmish/Elmish.WPF/blob/c10be058b14a4049ecfd8843f497bdfec1cf2c4a/src/Elmish.WPF/Binding.fs#L2334-L2337

In each case, model and message mapping can be done separately with functions like `Binding.mapModel` and `Binding.mapMsg` or `Binding.mapMsgWithModel`.
